core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -943,4 +955,71 @@ public String toString() {
       return subchannel.getLogId().toString();
     }
   }
+
+  @VisibleForTesting
+  static final class ManagedChannelPhantom extends PhantomReference<ManagedChannelImpl> {
+    private static final ReferenceQueue<ManagedChannelImpl> refQueue = initQueue();
+    // Retain the PhantomReferences so they don't get GC'd
```

Comment:
Just to be clear, this means that we probably won't detect problems in Servlet shutdown or equivalent, where the ClassLoader itself is being GC'd, right?

Reply:
Technically yes.  This is a best effort.   Even then, the phantoms don't do any clean up themselves they just act as a final notice that something wrong has happened.  EJ#7 suggests that finalizers may be used legitimately in this manner.